### ISSUE TITLE
About the training speed

### ISSUE THREAD

**Issue by youngzw** [THUwangcy/ReChorus#29]
Hi THUwangcy.

I use your lib to run the SASRec model with command as below (cuda environment):
```python
python main.py --model_name SASRec --emb_size 50 --lr 0.001 --l2 0.0 --dataset ml-1m --test_all 1 --history_max 200 --num_layers 2 --num_heads 1 --batch_size 128 --topk 10 --num_workers 2
```
The parameters are same with the original paper [SASRec](https://github.com/kang205/SASRec).  But I find the training speed is much slower than [the original code](https://github.com/kang205/SASRec)  for one epoch, even I modify the code to avoid inferring at every epoch.
https://github.com/THUwangcy/ReChorus/blob/9a4a783de1fdd02292fb95bc3471b8d310d2110a/src/helpers/BaseRunner.py#L121-L126
Can you provide some solutions to this problem?

**Comment by youngzw** [comment]
By the way, I wonder why the performance of SASRec model  is higher than the original paper. Here is the result with `ml-1m` dataset:
- [the original code](https://github.com/kang205/SASRec)  (I re-run it):
  - sample 99 negative items:
![image](https://user-images.githubusercontent.com/37242403/147734403-0390f338-eb5c-47ac-8848-b3b5a15b0052.png)

  - sample 100 negative items:
![image](https://user-images.githubusercontent.com/37242403/147734381-f7d0118b-fd0b-49e3-816e-5c017a92031d.png)

- screenshot of the results reported in the original paper (sample 100 negative items):
![image](https://user-images.githubusercontent.com/37242403/147734082-719eacba-85ef-4538-8cdb-56a784910558.png)


`rechorus` (default sample 99 negative items):
![image](https://user-images.githubusercontent.com/37242403/147734504-aeb19b6d-c0b4-4d9f-89f4-488a9719a66f.png)

**Comment by youngzw** [comment]
when ranking over all the items with `--test_all 1`:
- the original code (I modify it):
![image](https://user-images.githubusercontent.com/37242403/147735275-15f91cee-f333-46c6-b43b-0db671f52ea4.png)

- `rechorus`:
![image](https://user-images.githubusercontent.com/37242403/147735233-c3606f01-8815-495a-9f86-6d896e6bdf6d.png)

**Comment by THUwangcy** [comment]
I checked the original code of SASRec and find it adopts a quite different training paradigm. In our framework, a sequence with length 200 will be fed into the forward function 199 times. Each time the input includes the target item and corresponding history sequence. This is easier to understand and flexible to design more complex models (similar implementations can be found in [RecBole](https://github.com/RUCAIBox/RecBole/blob/master/recbole/model/sequential_recommender/sasrec.py)). However, the sequence will only be encoded once in the original code of SASRec. It will generate 200 logits corresponding to each position and use 199 of them to calculate the loss. This is far more efficient but requires the model to be able to output all the logits simultaneously. The difference in training paradigm might also lead to inconsistent performance.
